pkg/driver/controller.go

```diff
@@ -307,6 +314,14 @@ func (d *controllerService) ControllerPublishVolume(ctx context.Context, req *cs
 		return nil, status.Error(codes.InvalidArgument, errString)
 	}
 
+	// check if a request is already in-flight, VolumeId should be enough at this moment because Multi node writer mode is not support in the driver
+	// However, use req hash as key because eventually we need to support the ability to publish the same volume onto different nodes concurrently, and I don't see any harm here.
+	if ok := d.inFlight.Insert(req.String()); !ok {
```

Comment:
As we only support single writer mode now, I would say it is ok to just use volumeId as key here. However, I don't see any harm if we using combination of volumeId&nodeId, If customer is trying to attach same volume to different node now, this request will be reject in other steps. Advise?

Reply:
I removed this block in ControllerPublishVolume and ControllerUnpublishVolume, e2e test is failing because of this. I think we need to dig deeper into the publish/unpublish volume scenario. How about we implement this in a separate PR?

Reply:
which test? I agree, let's leave ControllerPublish/Attach for another PR, it's more complicated for sure

Reply:
the spec says: This operation MUST be idempotent. If the volume corresponding to the volume_id has already been published at the node corresponding to the node_id, and is compatible with the specified volume_capability and readonly flag, the Plugin MUST reply 0 OK.

So you are right, we should use volume_id and node_id.

If the attacher requests the same volume_id and different node_id (and volume is not multiattach, which we don't support yet anyway) the request will fail/get rejected anyway, I agree that's fine.

Reply:
Failed test: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_aws-ebs-csi-driver/924/pull-aws-ebs-csi-driver-e2e-single-az/1402061105047539712

Reply:
it seems like detach just took too long? or somehow the lock/inflight didn't get released, although the code is simple so I don't know how that would be possible.
```
 I0608 00:51:24.912672       1 cloud.go:594] Waiting for volume "vol-059485ebcfd5f7282" state: actual=detaching, desired=detached
I0608 00:51:26.006673       1 cloud.go:594] Waiting for volume "vol-059485ebcfd5f7282" state: actual=detaching, desired=detached
I0608 00:51:27.918180       1 cloud.go:594] Waiting for volume "vol-059485ebcfd5f7282" state: actual=detaching, desired=detached
W0608 00:53:00.254490       1 cloud.go:535] Ignoring error from describe volume for volume "vol-0944244f08b5badd0"; will retry: "RequestCanceled: request context canceled\ncaused by: context deadline exceeded"
E0608 00:53:28.688453       1 driver.go:119] GRPC error: rpc error: code = Aborted desc = ControllerUnpublishVolume for Volume vol-0944244f08b5badd0 and Node i-071ebf53811520106 is already in progress
E0608 00:54:18.254315       1 driver.go:119] GRPC error: rpc error: code = Aborted desc = ControllerUnpublishVolume for Volume vol-0944244f08b5badd0 and Node i-071ebf53811520106 is already in progress 
```

Reply:
I wonder if it's possible for attach/detach to deadlock since in kubernetes world it's quite common/possible to have 1 replica Terminating on one Node and its vol trying to detach at the same time as 1 replica Pending on another Node and its vol trying to attach. This would mean we would need to index by ControllerPublish+volume_id+node_id or ontrollerUnpublish+volume_id+node_id